docs/docs/reference/metaprogramming/macros.md

```diff
@@ -564,6 +564,53 @@ while (i < arr.length) {
 }
 sum
 ```
+### Showing meaningful definition names in quotes
+
+In the `powerCode` example above there is a `'{ val y = $x * $x; ... }` which when printed
+may show several different `val y = ...`.
+
+For example
+```scala
+powerCode(16, '{7}).show
+```
+will show
+```scala
+val y: scala.Double = 7
+val y: scala.Double = y.*(y)
+val y: scala.Double = y.*(y)
+val y: scala.Double = y.*(y)
+val y: scala.Double = y.*(y)
+y
+```
+Even though there is no hygiene issue it may be hard to undestand the code. To overcome this inconvenience
+each `y` can be assigned a meaningful name using the `scala.quoted.show.showName` annotation.
+For example `'{ @showName(${Expr("y" + i)}) val y = $x * $x; ... }` will assign to each `y` a name
+`a{i}` where `{i}` is a known String, if `i == 3` then it would be named `a3`.
+
+The `powerCode` can be defined as follows using `showName`
+```scala
+def powerCodeD(n: Long, x: Expr[Double]))(given QuoteContext): Expr[Double] = '{
+  val a1 = $x
+  ${ powerCodeD(n, 2, 'x1) }
+}
+def powerCodeD(n: Long, i: Int, x: Expr[Double])(given QuoteContext): Expr[Double] =
+  if (n == 0) '{1.0}
+  else if (n % 2 == 0) '{ @showName(${Expr("a" + i)}) val y = $x * $x; ${powerCodeD(n / 2, idx * 2, 'y)} }
```

Comment:
Should we instead provide a means to developers for generating fresh identifiers?

```suggestion
  else if (n % 2 == 0) {
    val y = freshName()
    '{ val $y = $x * $x; ${powerCodeD(n / 2, idx * 2, y)} }
  }
```

Also, I’m very surprised that there is no name clash in the expanded code you shown. It seems that all the lines of the block are in the same lexical scope.

Reply:
Actually, I’m not sure about the suggestion above. It can not express something like `'{ val MyExtractor(y) = $x * $x }`, for instance.
Otherwise, could we support something like the following?

```suggestion
  else if (n % 2 == 0) {
    val y = name("a" + i)
    '{ val $y = $x * $x; ${powerCodeD(n / 2, idx * 2, y)} }
  }
```

I believe we can’t, though…

Reply:
The annotation is the first step towards splicing a name. This PR is a starting point to support all those patterns. Ideally a `'{ val $y = ... }` will be desugared to `'{ @showName(${Expr(y.toString)}) val y = ... }`